lnwallet/musig_session.go

```diff
@@ -0,0 +1,535 @@
+// them to generate another signature.
+func (m *MusigSession) VerifyCommitSig(commitTx *wire.MsgTx,
+	sig *lnwire.PartialSigWithNonce,
+	musigOpts ...MusigSessionOpt) (*musig2.Nonces, error) {
+
+	opts := defaultMusigSessionOpts()
+	for _, optFunc := range musigOpts {
+		optFunc(opts)
+	}
+
+	// Before we can verify the signature, we'll need to finalize the
+	// session by binding the remote party's provided signing nonce.
+	if err := m.FinalizeSession(musig2.Nonces{
+		PubNonce: sig.Nonce,
+	}); err != nil {
+		return nil, err
+	}
+
+	// When we verify a commitment signature, we always assume that we're
+	// verifying a signature on our local commitment. Therefore, we'll use:
+	// their remote nonce, and also public key.
+	partialSig := NewMusigPartialSig(
+		&musig2.PartialSignature{S: &sig.Sig},
+		m.nonces.SigningNonce.PubNonce, m.combinedNonce, m.signerKeys,
+	)
+
+	walletLog.Infof("verify partial sig: %v", spew.Sdump(partialSig))
+
+	// With the partial sig loaded with the proper context, we'll now
+	// generate the sighash that the remote party should have signed.
+	sigHash, err := taprootKeyspendSighash(
+		commitTx, m.inputTxOut.PkScript, m.inputTxOut.Value,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	walletLog.Infof("Verfiying new musig2 sig for session=%x, nonce=%s",
+		m.session.SessionID[:], m.nonces.String())
+
+	if !partialSig.Verify(sigHash, m.remoteKey.PubKey) {
+		return nil, fmt.Errorf("invalid partial commit sig")
+	}
+
+	nonceOpts := []musig2.NonceGenOption{
+		musig2.WithPublicKey(m.localKey.PubKey),
+	}
+	if opts.customRand != nil {
+		nonceOpts = append(
+			nonceOpts, musig2.WithCustomRand(opts.customRand),
+		)
+	}
+
+	// At this point, we know that their signature is valid, so we'll
+	// generate another verification nonce for them, so they can generate a
+	// new state transition.
+	nextVerificationNonce, err := musig2.GenNonces(nonceOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("unable to gen new nonce: %w", err)
+	}
+
+	return nextVerificationNonce, nil
+}
+
+// CombineSigs combines the passed partial signatures into a valid schnorr
+// signature.
+func (m *MusigSession) CombineSigs(sigs ...*musig2.PartialSignature,
+) (*schnorr.Signature, error) {
+
+	defer m.signer.MuSig2Cleanup(m.session.SessionID)
+
+	sig, _, err := m.signer.MuSig2CombineSig(
+		m.session.SessionID,
+		sigs,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return sig, nil
+}
+
+// MusigSessionCfg is used to create a new musig2 pair session. It contains the
+// keys for both parties, as well as their initial verification nonces.
+type MusigSessionCfg struct {
+	// LocalKey is a key desc for the local key.
+	LocalKey keychain.KeyDescriptor
+
+	// RemoteKey is a key desc for the remote key.
+	RemoteKey keychain.KeyDescriptor
+
+	// LocalNonce is the local party's initial verification nonce.
+	LocalNonce musig2.Nonces
+
+	// RemoteNonce is the remote party's initial verification nonce.
```

Comment:
and ie, the local party's signing nonce right?

Reply:
Correcto, I don't think the terms here are amazing, but I think it's clearer that it was before. The code is also currently using half signing/verification and half local/remote.